docs/concepts/caching.rst

```diff
@@ -0,0 +1,489 @@
+
+
+.. code-block:: python
+
+    from hamilton import driver
+    import my_dataflow
+
+    dr = (
+        driver.Builder()
+        .with_module(my_dataflow)
+        .with_cache()
+        .build()
+    )
+
+    dr.execute([...])
+
+
+Cache to a file
+-----------------
+
+By default, caching only uses the Python standard library and relies on the ``pickle`` format to store results, `which comes with caveats <https://grantjenks.com/docs/diskcache/tutorial.html#caveats>`_ and can be impractical. Furthermore, some Python objects can't even be pickled (e.g., text tokenizers with Rust bindings).
+
+Using the Hamilton ``@tag`` decorator, you can cache a node's result to a specific file format (``JSON``, ``CSV``, ``Parquet``, etc.) and circumvent the limitations of ``pickle`` files.
+
+
+.. code-block:: python
+
+    # my_dataflow.py
+    import pandas as pd
+    from hamilton.function_modifiers import tag
+
+    def raw_data(path: str) -> pd.DataFrame:
+        return pd.read_csv(path)
+
+    @tag(cache="parquet")
+    def clean_dataset(raw_data: pd.DataFrame) -> pd.DataFrame:
+        raw_data = raw_data.fillna(0)
+        return raw_data
+
+    @tag(cache="json")
+    def statistics(clean_dataset: pd.DataFrame) -> dict:
+        return ...
+
+
+.. note::
+
+    This aims to match and replace the features of ``hamilton.experimental.h_cache.CachingGraphAdater`` but the API is preliminary.
+
+
+Resume from checkpoint
+-----------------------
+
+The caching feature stores results after each node execution. If execution fails half-way, the ``Driver`` can use the cache metadata and results to resume the run. You can think of it as automatically passing ``.execute(overrides={...})`` from disk.
+
+To resume, you can specify ``.with_cache(resume_from=...)`` which accepts a specific ``run_id`` or the string ``"latest"`` which will automatically get the results from the latest execution, whether it succeeded or failed. The ``run_id`` is automatically generated internally and can be viewed via the cache logger.
+
+.. code-block:: python
+
+    from hamilton import driver
+    import my_dataflow
+
+    dr = (
+        driver.Builder()
+        .with_modules(my_dataflow)
+        .with_cache(resume_from="latest")
+        .build()
+    )
+
+    dr.execute(...)
+
+.. note::
+
+    Using ``resume_from="latest"`` gets the latest run at the moment of building the ``Driver``. Consequently, calling ``Driver.execute()`` multiple times afterwards will always retrieve the same results from that specific point in time.
+
+
```

Comment:
There needs to be a  "Only recompute what has changed" section to complement the above "checkpointing section".

E.g. 
2. run dataflow with caching.
3. there is nothing in the cache -- so compute everything.
4. it runs through to completion
5. we then run the code again but change an input
6. we only recompute the parts of the dataflow impacted by that new data
7. we then run the code again, but change code
8. we only recompute the parts of the dataflow impacted by that code change

Reply:
What you describe as "only recompute what has changed" is the default behavior described under "Basic". Unless you mean something else